### PR TITLE
Refactor SparseApply[Proximal]Adagrad into classes

### PR DESCRIPTION
**Pull request by benbarsdell**
- This is a NFC in preparation for adding GPU implementations.
- It also combines the SparseApplyAdagrad[V2] kernels into a single implementation distinguished by a has_epsilon template parameter.

cc @nluehr

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.h
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.h:100]: Please document what is the semantic of `epsilon` if `has_epsilon` is false (it is ignored right?).

File: tensorflow/core/kernels/training_ops.cc
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:213]: Stray newline?

File: tensorflow/core/kernels/training_ops.cc
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:177]: Please use early exit to reduce indentation.

File: tensorflow/core/kernels/training_ops.cc
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:292]: Please use early exit to reduce indentation.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Only minor comments (I'm assuming this is just moving code around).

File: tensorflow/core/kernels/training_ops.h
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.h:101]: Done.

File: tensorflow/core/kernels/training_ops.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc:221]: Fixed.

File: tensorflow/core/kernels/training_ops.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc]: Done.

File: tensorflow/core/kernels/training_ops.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc]: Done.

**Comment by irvineoy** [comment]
Sorry for commenting on this PR. I implemented the GPU kernel of ResourceSparseApplyAdadelta which is still waiting for review (#44347). And I found Adagrad optimizer will also encounter the same problem when dealing with sparse tensor on GPU. So I wonder do you consider implementing the GPU kernel of ResourceSparseApplyAdagrad as a good way to solve it and if this a good part to contribute? Thank you very much. @sanjoy @benbarsdell

**Comment by benbarsdell** [comment]
@irvineoy thanks for your interest. I have implemented a GPU version of the ops in this PR and will be submitting it in a follow-up PR as soon as this one is merged.

**Review by sanjoy — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**
